ReSharper.FSharp/src/FSharp/FSharp.Psi.Intentions/src/Intentions/FunctionAnnotationAction.fs

```diff
@@ -35,6 +35,60 @@ module SpecifyTypes =
         static member ParameterTypesOnly = { CanSpecifyParameterTypes = true; CanSpecifyReturnType = false }
         static member ReturnTypeOnly = { CanSpecifyParameterTypes = false; CanSpecifyReturnType = true }
 
+    let rec private collectTypeUsages acc (context, fcsType: FSharpType, typeUsage: ITypeUsage) =
+        if fcsType.IsGenericParameter then acc else
+
+        match typeUsage with
+        | :? INamedTypeUsage as typeUsage ->
+            let typeReference = typeUsage.ReferenceName
+            let acc = (typeUsage, fcsType, context) :: acc
+
+            let typeArgs = typeReference.TypeArgumentList
+            if isNull typeArgs then acc else
+
+            typeArgs.TypeUsagesEnumerable
+            |> Seq.mapi (fun i typeArg -> i, typeArg)
+            |> Seq.fold (fun acc (i, typeArg) ->
+                let fcsType = fcsType.GenericArguments[i]
```

Comment:
What happens if there's a different count of type arguments between the syntax and inferred by FCS?
Let's assume, the type checker can recover incorrect count in the syntax.

Reply:
Can you show me an example? We get ITypeUsage exactly from FSharpType, in which cases can we have a difference between the syntax and FCS inferring?